### DISCUSSION TITLE
Why weren't tuples implemented using anonymous types?

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5114] [Language Ideas]
The same mechanism can be applied when the scope is a class or even a project (assembly). Between assemblies, a duplicate type can be made. If project B relies on project A, and A has anonymous type X, and B needs to interact with a member of A which deals with X, then a duplicate type Y is created in B, and it defines implicit conversion operators to/from X.

**Comment by HaloFour** [comment]
Anonymous types are unspeakable (thus cannot be exposed publicly) immutable reference types which require heap allocations.  The desire was to have tuples be sharable across assemblies, mutable values types.

**Reply by TahirAhmadov** [reply]
I described in my question how anonymous types can be made "speakable". True, tuples are mutable, but a) mutability of tuples isn't often used - in the typical scenario of returning a complex value, you construct it in the method and the calling site deconstructs it immediately; b) mutability goes directly against the overall shift to functional programming; c) there is no reason why anonymous types had to stay immutable - if it's OK for tuples, why is it not OK for anonymous types? especially given that a new syntax is created for "tuples" (the construct, not the implementation); `(123, "abc")` becomes an anonymous mutable struct in one shot. 

Tuples have disadvantages. They implicitly convert way to easily to other tuples; they introduced a new way of creating nameless types whereas anonymous types already existed; and anonymous types have actual members when looking at them through reflection (granted, I'm not sure how this works for ValueTuple, I imagine you have to use attributes and translate from member names to `Item1` and stuff?)

```cs
(int A, string B) Foo((DateTime C, TimeSpan D) x)
{
  var (c, d) = x; // deconstruct; all the tuple syntax can work with anonymous structs
  return (123, "abc");
}
// This compiles into 2 types of the same access as the method.

// another project
void Bar()
{
  var (a,b) = Project1.Class.Foo((DateTime.Now, TimeSpan.Zero));
  // 2 new types are created; they have implicit operators defined to convert from Project1's version to our local version
}
```

**Reply by HaloFour** [reply]
The debate around this design has already happened and the feature has already shipped, it's much too late to reconsider it.  All of your points, however, were brought up during the design discussions and addressed.

> b) mutability goes directly against the overall shift to functional programming; c) there is no reason why anonymous types had to stay immutable - if it's OK for tuples, why is it not OK for anonymous types?

Mutability is a problem when combined with sharing.  Anonymous types are shared references, thus mutability would be a problem.  Value types are copied thus mutations are not observable, so these concerns are generally moot.

> They implicitly convert way to easily to other tuples; they introduced a new way of creating nameless types whereas anonymous types already existed; and anonymous types have actual members when looking at them through reflection

This is intentional.  The names exist for the benefit of the developer, they are not a part of the contract of the type.  They have about the same value as the names of parameters.

**Reply by MgSam** [reply]
> Anonymous types are unspeakable...

I prefer to call them Types-Who-Must-Not-Be-Named

**Reply by Joe4evr** [reply]
> I prefer to call them Types-Who-Must-Not-Be-Named

You jest, [but D actually has that](https://wiki.dlang.org/Voldemort_types).

**Reply by CyrusNajmabadi** [reply]
> there is no reason why anonymous types had to stay immutable

That would definitely be undesirable to change.  They would lose their ability to safely act as values of they could mutate.
